{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "# Viewing & validating the mapping file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## File location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-14T18:46:23.960883Z",
     "start_time": "2017-09-14T18:46:23.830919Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 1.7M\r\n",
      "drwxr-xr-x 2 nyoungblut ebio  132 Sep 14 20:46 \u001b[0m\u001b[01;34msplit_libraries\u001b[0m/\r\n",
      "-rw-r--r-- 1 nyoungblut ebio 118K Sep 14 20:46 rep_set.tre\r\n",
      "-rw-r--r-- 1 nyoungblut ebio 542K Sep 14 20:46 rep_set.fna\r\n",
      "-rw-r--r-- 1 nyoungblut ebio 1.1M Sep 14 20:46 otu_table.biom\r\n",
      "-rw-r--r-- 1 nyoungblut ebio 2.9K Sep 14 20:46 qiime_processing_workflow_local.sh\r\n",
      "-rw------- 1 nyoungblut ebio 3.0K Sep 14 20:46 mapping_file.txt\r\n",
      "-rw-r--r-- 1 nyoungblut ebio  108 Sep 14 20:46 16s_pickotu_param.txt\r\n"
     ]
    }
   ],
   "source": [
    "ls -thlc ../../data/Qiime_Introduction_Tutorial/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-14T18:46:24.956486Z",
     "start_time": "2017-09-14T18:46:24.947007Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/ebio/abt3_projects/software/tmp/data_analysis_workshops/data/Qiime_Introduction_Tutorial\n"
     ]
    }
   ],
   "source": [
    "%cd ../../data/Qiime_Introduction_Tutorial/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-14T18:46:25.936528Z",
     "start_time": "2017-09-14T18:46:25.815386Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 1.7M\r\n",
      "drwxr-xr-x 2 nyoungblut ebio  132 Sep 14 20:46 \u001b[0m\u001b[01;34msplit_libraries\u001b[0m/\r\n",
      "-rw-r--r-- 1 nyoungblut ebio 118K Sep 14 20:46 rep_set.tre\r\n",
      "-rw-r--r-- 1 nyoungblut ebio 542K Sep 14 20:46 rep_set.fna\r\n",
      "-rw-r--r-- 1 nyoungblut ebio 1.1M Sep 14 20:46 otu_table.biom\r\n",
      "-rw-r--r-- 1 nyoungblut ebio 2.9K Sep 14 20:46 qiime_processing_workflow_local.sh\r\n",
      "-rw------- 1 nyoungblut ebio 3.0K Sep 14 20:46 mapping_file.txt\r\n",
      "-rw-r--r-- 1 nyoungblut ebio  108 Sep 14 20:46 16s_pickotu_param.txt\r\n"
     ]
    }
   ],
   "source": [
    "ls -thlc "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Loading file with pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-14T18:46:29.808799Z",
     "start_time": "2017-09-14T18:46:27.713869Z"
    },
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-14T18:46:29.880435Z",
     "start_time": "2017-09-14T18:46:29.811268Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>SampleType</th>\n",
       "      <th>Year</th>\n",
       "      <th>Month</th>\n",
       "      <th>Day</th>\n",
       "      <th>Subject</th>\n",
       "      <th>ReportedAntibioticUsage</th>\n",
       "      <th>DaysSinceExperimentStart</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>L1S8</td>\n",
       "      <td>AGCTGACTAGTC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>gut</td>\n",
       "      <td>2008</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0</td>\n",
       "      <td>1_Fece_10_28_2008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1S140</td>\n",
       "      <td>ATGGCAGCTCTA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>gut</td>\n",
       "      <td>2008</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0</td>\n",
       "      <td>2_Fece_10_28_2008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L1S57</td>\n",
       "      <td>ACACACTATGGC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>gut</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>84</td>\n",
       "      <td>1_Fece_1_20_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L1S208</td>\n",
       "      <td>CTGAGATACGCG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>gut</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>84</td>\n",
       "      <td>2_Fece_1_20_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L1S76</td>\n",
       "      <td>ACTACGTGTGGT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>gut</td>\n",
       "      <td>2009</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>112</td>\n",
       "      <td>1_Fece_2_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>L1S105</td>\n",
       "      <td>AGTGCGATGCGT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>gut</td>\n",
       "      <td>2009</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>140</td>\n",
       "      <td>1_Fece_3_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>L1S257</td>\n",
       "      <td>CCGACTGAGATG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>gut</td>\n",
       "      <td>2009</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>140</td>\n",
       "      <td>2_Fece_3_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>L1S281</td>\n",
       "      <td>CCTCTCGTGATC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>gut</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>168</td>\n",
       "      <td>2_Fece_4_14_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>L2S240</td>\n",
       "      <td>CATATCGCAGTT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>left palm</td>\n",
       "      <td>2008</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0</td>\n",
       "      <td>2_L_Palm_10_28_2008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>L2S155</td>\n",
       "      <td>ACGATGCGACCA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>left palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>84</td>\n",
       "      <td>1_L_Palm_1_20_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>L2S309</td>\n",
       "      <td>CGTGCATTATCA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>left palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>84</td>\n",
       "      <td>2_L_Palm_1_20_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>L2S175</td>\n",
       "      <td>AGCTATCCACGA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>left palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>112</td>\n",
       "      <td>1_L_Palm_2_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>L2S204</td>\n",
       "      <td>ATGCAGCTCAGT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>left palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>140</td>\n",
       "      <td>1_L_Palm_3_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>L2S357</td>\n",
       "      <td>CTAACGCAGTCA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>left palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>140</td>\n",
       "      <td>2_L_Palm_3_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>L2S222</td>\n",
       "      <td>CACGTGACATGT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>left palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>168</td>\n",
       "      <td>1_L_Palm_4_14_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>L2S382</td>\n",
       "      <td>CTCAATGACTCA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>left palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>168</td>\n",
       "      <td>2_L_Palm_4_14_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>L3S242</td>\n",
       "      <td>ACAGTTGCGCGA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>right palm</td>\n",
       "      <td>2008</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0</td>\n",
       "      <td>1_R_Palm_10_28_2008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>L3S378</td>\n",
       "      <td>ATCGATCTGTGG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>right palm</td>\n",
       "      <td>2008</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0</td>\n",
       "      <td>2_R_Palm_10_28_2008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>L3S294</td>\n",
       "      <td>CACGACAGGCTA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>right palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>84</td>\n",
       "      <td>1_R_Palm_1_20_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>L4S63</td>\n",
       "      <td>CTCGTGGAGTAG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>right palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>84</td>\n",
       "      <td>2_R_Palm_1_20_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>L3S313</td>\n",
       "      <td>AGTGTCACGGTG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>right palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>112</td>\n",
       "      <td>1_R_Palm_2_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>L3S341</td>\n",
       "      <td>CAAGTGAGAGAG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>right palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>140</td>\n",
       "      <td>1_R_Palm_3_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>L4S112</td>\n",
       "      <td>GCGTTACACACA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>right palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>140</td>\n",
       "      <td>2_R_Palm_3_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>L3S360</td>\n",
       "      <td>CATCGTATCAAC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>right palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>168</td>\n",
       "      <td>1_R_Palm_4_14_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>L4S137</td>\n",
       "      <td>GAACTGTATCTC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>right palm</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>168</td>\n",
       "      <td>2_R_Palm_4_14_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>L5S104</td>\n",
       "      <td>CAGTGTCAGGAC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>tongue</td>\n",
       "      <td>2008</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0</td>\n",
       "      <td>1_Tong_10_28_2008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>L5S240</td>\n",
       "      <td>CTGGACTCATAG</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>tongue</td>\n",
       "      <td>2008</td>\n",
       "      <td>10</td>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>Yes</td>\n",
       "      <td>0</td>\n",
       "      <td>2_Tong_10_28_2008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>L5S155</td>\n",
       "      <td>ATCTTAGACTGC</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>tongue</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>84</td>\n",
       "      <td>1_Tong_1_20_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>L6S20</td>\n",
       "      <td>GAGGCTCATCAT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>tongue</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>84</td>\n",
       "      <td>2_Tong_1_20_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>L5S174</td>\n",
       "      <td>CAGACATTGCGT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>tongue</td>\n",
       "      <td>2009</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>112</td>\n",
       "      <td>1_Tong_2_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>L5S203</td>\n",
       "      <td>CGATGCACCAGA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>tongue</td>\n",
       "      <td>2009</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>140</td>\n",
       "      <td>1_Tong_3_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>L6S68</td>\n",
       "      <td>GATACGTCCTGA</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>tongue</td>\n",
       "      <td>2009</td>\n",
       "      <td>3</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>140</td>\n",
       "      <td>2_Tong_3_17_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>L5S222</td>\n",
       "      <td>CTAGAGACTCTT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>tongue</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>168</td>\n",
       "      <td>1_Tong_4_14_2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>L6S93</td>\n",
       "      <td>GATTAGCACTCT</td>\n",
       "      <td>GTGCCAGCMGCCGCGGTAA</td>\n",
       "      <td>tongue</td>\n",
       "      <td>2009</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>168</td>\n",
       "      <td>2_Tong_4_14_2009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #SampleID BarcodeSequence LinkerPrimerSequence  SampleType  Year  Month  \\\n",
       "0       L1S8    AGCTGACTAGTC  GTGCCAGCMGCCGCGGTAA         gut  2008     10   \n",
       "1     L1S140    ATGGCAGCTCTA  GTGCCAGCMGCCGCGGTAA         gut  2008     10   \n",
       "2      L1S57    ACACACTATGGC  GTGCCAGCMGCCGCGGTAA         gut  2009      1   \n",
       "3     L1S208    CTGAGATACGCG  GTGCCAGCMGCCGCGGTAA         gut  2009      1   \n",
       "4      L1S76    ACTACGTGTGGT  GTGCCAGCMGCCGCGGTAA         gut  2009      2   \n",
       "5     L1S105    AGTGCGATGCGT  GTGCCAGCMGCCGCGGTAA         gut  2009      3   \n",
       "6     L1S257    CCGACTGAGATG  GTGCCAGCMGCCGCGGTAA         gut  2009      3   \n",
       "7     L1S281    CCTCTCGTGATC  GTGCCAGCMGCCGCGGTAA         gut  2009      4   \n",
       "8     L2S240    CATATCGCAGTT  GTGCCAGCMGCCGCGGTAA   left palm  2008     10   \n",
       "9     L2S155    ACGATGCGACCA  GTGCCAGCMGCCGCGGTAA   left palm  2009      1   \n",
       "10    L2S309    CGTGCATTATCA  GTGCCAGCMGCCGCGGTAA   left palm  2009      1   \n",
       "11    L2S175    AGCTATCCACGA  GTGCCAGCMGCCGCGGTAA   left palm  2009      2   \n",
       "12    L2S204    ATGCAGCTCAGT  GTGCCAGCMGCCGCGGTAA   left palm  2009      3   \n",
       "13    L2S357    CTAACGCAGTCA  GTGCCAGCMGCCGCGGTAA   left palm  2009      3   \n",
       "14    L2S222    CACGTGACATGT  GTGCCAGCMGCCGCGGTAA   left palm  2009      4   \n",
       "15    L2S382    CTCAATGACTCA  GTGCCAGCMGCCGCGGTAA   left palm  2009      4   \n",
       "16    L3S242    ACAGTTGCGCGA  GTGCCAGCMGCCGCGGTAA  right palm  2008     10   \n",
       "17    L3S378    ATCGATCTGTGG  GTGCCAGCMGCCGCGGTAA  right palm  2008     10   \n",
       "18    L3S294    CACGACAGGCTA  GTGCCAGCMGCCGCGGTAA  right palm  2009      1   \n",
       "19     L4S63    CTCGTGGAGTAG  GTGCCAGCMGCCGCGGTAA  right palm  2009      1   \n",
       "20    L3S313    AGTGTCACGGTG  GTGCCAGCMGCCGCGGTAA  right palm  2009      2   \n",
       "21    L3S341    CAAGTGAGAGAG  GTGCCAGCMGCCGCGGTAA  right palm  2009      3   \n",
       "22    L4S112    GCGTTACACACA  GTGCCAGCMGCCGCGGTAA  right palm  2009      3   \n",
       "23    L3S360    CATCGTATCAAC  GTGCCAGCMGCCGCGGTAA  right palm  2009      4   \n",
       "24    L4S137    GAACTGTATCTC  GTGCCAGCMGCCGCGGTAA  right palm  2009      4   \n",
       "25    L5S104    CAGTGTCAGGAC  GTGCCAGCMGCCGCGGTAA      tongue  2008     10   \n",
       "26    L5S240    CTGGACTCATAG  GTGCCAGCMGCCGCGGTAA      tongue  2008     10   \n",
       "27    L5S155    ATCTTAGACTGC  GTGCCAGCMGCCGCGGTAA      tongue  2009      1   \n",
       "28     L6S20    GAGGCTCATCAT  GTGCCAGCMGCCGCGGTAA      tongue  2009      1   \n",
       "29    L5S174    CAGACATTGCGT  GTGCCAGCMGCCGCGGTAA      tongue  2009      2   \n",
       "30    L5S203    CGATGCACCAGA  GTGCCAGCMGCCGCGGTAA      tongue  2009      3   \n",
       "31     L6S68    GATACGTCCTGA  GTGCCAGCMGCCGCGGTAA      tongue  2009      3   \n",
       "32    L5S222    CTAGAGACTCTT  GTGCCAGCMGCCGCGGTAA      tongue  2009      4   \n",
       "33     L6S93    GATTAGCACTCT  GTGCCAGCMGCCGCGGTAA      tongue  2009      4   \n",
       "\n",
       "    Day  Subject ReportedAntibioticUsage  DaysSinceExperimentStart  \\\n",
       "0    28        1                     Yes                         0   \n",
       "1    28        2                     Yes                         0   \n",
       "2    20        1                      No                        84   \n",
       "3    20        2                      No                        84   \n",
       "4    17        1                      No                       112   \n",
       "5    17        1                      No                       140   \n",
       "6    17        2                      No                       140   \n",
       "7    14        2                      No                       168   \n",
       "8    28        2                     Yes                         0   \n",
       "9    20        1                      No                        84   \n",
       "10   20        2                      No                        84   \n",
       "11   17        1                      No                       112   \n",
       "12   17        1                      No                       140   \n",
       "13   17        2                      No                       140   \n",
       "14   14        1                      No                       168   \n",
       "15   14        2                      No                       168   \n",
       "16   28        1                     Yes                         0   \n",
       "17   28        2                     Yes                         0   \n",
       "18   20        1                      No                        84   \n",
       "19   20        2                      No                        84   \n",
       "20   17        1                      No                       112   \n",
       "21   17        1                      No                       140   \n",
       "22   17        2                      No                       140   \n",
       "23   14        1                      No                       168   \n",
       "24   14        2                      No                       168   \n",
       "25   28        1                     Yes                         0   \n",
       "26   28        2                     Yes                         0   \n",
       "27   20        1                      No                        84   \n",
       "28   20        2                      No                        84   \n",
       "29   17        1                      No                       112   \n",
       "30   17        1                      No                       140   \n",
       "31   17        2                      No                       140   \n",
       "32   14        1                      No                       168   \n",
       "33   14        2                      No                       168   \n",
       "\n",
       "            Description  \n",
       "0     1_Fece_10_28_2008  \n",
       "1     2_Fece_10_28_2008  \n",
       "2      1_Fece_1_20_2009  \n",
       "3      2_Fece_1_20_2009  \n",
       "4      1_Fece_2_17_2009  \n",
       "5      1_Fece_3_17_2009  \n",
       "6      2_Fece_3_17_2009  \n",
       "7      2_Fece_4_14_2009  \n",
       "8   2_L_Palm_10_28_2008  \n",
       "9    1_L_Palm_1_20_2009  \n",
       "10   2_L_Palm_1_20_2009  \n",
       "11   1_L_Palm_2_17_2009  \n",
       "12   1_L_Palm_3_17_2009  \n",
       "13   2_L_Palm_3_17_2009  \n",
       "14   1_L_Palm_4_14_2009  \n",
       "15   2_L_Palm_4_14_2009  \n",
       "16  1_R_Palm_10_28_2008  \n",
       "17  2_R_Palm_10_28_2008  \n",
       "18   1_R_Palm_1_20_2009  \n",
       "19   2_R_Palm_1_20_2009  \n",
       "20   1_R_Palm_2_17_2009  \n",
       "21   1_R_Palm_3_17_2009  \n",
       "22   2_R_Palm_3_17_2009  \n",
       "23   1_R_Palm_4_14_2009  \n",
       "24   2_R_Palm_4_14_2009  \n",
       "25    1_Tong_10_28_2008  \n",
       "26    2_Tong_10_28_2008  \n",
       "27     1_Tong_1_20_2009  \n",
       "28     2_Tong_1_20_2009  \n",
       "29     1_Tong_2_17_2009  \n",
       "30     1_Tong_3_17_2009  \n",
       "31     2_Tong_3_17_2009  \n",
       "32     1_Tong_4_14_2009  \n",
       "33     2_Tong_4_14_2009  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('mapping_file.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Validating mapping file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Note: Python 3 is used by default for BASH, and QIIME is only Python 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-14T18:46:43.424630Z",
     "start_time": "2017-09-14T18:46:43.404642Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "bash: line 1: validate_mapping_file.py: command not found\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "validate_mapping_file.py -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2017-09-14T18:46:44.252630Z",
     "start_time": "2017-09-14T18:46:44.229441Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Python 3.6.1 :: Continuum Analytics, Inc.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# python2 or python3?\n",
    "python --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2017-09-14T18:46:48.180Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: validate_mapping_file.py [options] {-m/--mapping_fp MAPPING_FP}\n",
      "\n",
      "[] indicates optional input (order unimportant)\n",
      "{} indicates required input (order unimportant)\n",
      "\n",
      "Specifically, we check that:\n",
      "\n",
      "    - The BarcodeSequence, LinkerPrimerSequences, and ReversePrimer fields\n",
      "       have valid IUPAC DNA characters, and BarcodeSequence characters\n",
      "       are non-degenerate (error)\n",
      "    - The SampleID, BarcodeSequence, LinkerPrimerSequence, and Description\n",
      "       headers are present. (error)\n",
      "    - There are not duplicate header fields (error)\n",
      "    - There are not duplicate barcodes (error)\n",
      "    - Barcodes are of the same length.  Suppressed when\n",
      "       variable_len_barcode flag is passed (warning)\n",
      "    - The headers do not contain invalid characters (alphanumeric and\n",
      "       underscore only) (warning)\n",
      "    - The data fields do not contain invalid characters (alphanumeric,\n",
      "       underscore, space, and +-%./:,; characters) (warning)\n",
      "    - SampleID fields are MIENS compliant (only alphanumeric\n",
      "       and . characters). (warning)\n",
      "    - There are no duplicates when the primer and variable length\n",
      "       barcodes are appended (error)\n",
      "    - There are no duplicates when barcodes and added demultiplex\n",
      "       fields (-j option) are combined (error)\n",
      "    - Data fields are not found beyond the Description column (warning)\n",
      "\n",
      "    Details about the metadata mapping file format can be found here:\n",
      "    http://www.qiime.org/documentation/file_formats.html#metadata-mapping-files\n",
      "\n",
      "    Errors and warnings are saved to a log file.  Errors can be caused by\n",
      "    problems with the headers, invalid characters in barcodes or primers, or\n",
      "    by duplications in SampleIDs or barcodes.\n",
      "\n",
      "    Warnings can arise from invalid characters and variable length barcodes that\n",
      "    are not specified with the --variable_len_barcode.\n",
      "    Warnings will contain a reference to the cell (row,column) that the\n",
      "    warning arose from.\n",
      "\n",
      "    In addition to the log file, a \"corrected_mapping\" file will be created.\n",
      "    Any invalid characters will be replaced with '.' characters in\n",
      "    the SampleID fields (to enforce MIENS compliance) and text in other data\n",
      "    fields will be replaced with the character specified by the -c parameter,\n",
      "    which is an underscore \"_\" by default.\n",
      "\n",
      "    A html file will be created as well, which will show locations of\n",
      "    warnings and errors, highlighted in yellow and red respectively.  If no\n",
      "    errors or warnings were present the file will display a message saying\n",
      "    such.  Header errors can mask other errors, so these should be corrected\n",
      "    first.\n",
      "\n",
      "    If pooled primers are used, separate with a comma.  For instance, a pooled\n",
      "    set of three 27f primers (used to increase taxonomic coverage) could be\n",
      "    specified in the LinkerPrimerSequence fields as such:\n",
      "    AGGGTTCGATTCTGGCTCAG,AGAGTTTGATCCTGGCTTAG,AGAATTTGATCTTGGTTCAG\n",
      "\n",
      "\n",
      "Example usage: \n",
      "Print help message and exit\n",
      " validate_mapping_file.py -h\n",
      "\n",
      "Example: Check the Fasting_Map.txt     mapping file for problems, supplying the required mapping file, and output     the results in the validate_mapping_file_output directory\n",
      " validate_mapping_file.py -m     Fasting_Map.txt -o validate_mapping_file_output\n",
      "\n",
      "Options:\n",
      "  --version             show program's version number and exit\n",
      "  -h, --help            show this help message and exit\n",
      "  -o OUTPUT_DIR, --output_dir=OUTPUT_DIR\n",
      "                        Required output directory for log file, corrected\n",
      "                        mapping file, and html file. [default: ./]\n",
      "  -v, --verbose         Enable printing information to standard out [default:\n",
      "                        True]\n",
      "  -c CHAR_REPLACE, --char_replace=CHAR_REPLACE\n",
      "                        Changes the default character used to replace invalid\n",
      "                        characters found in the mapping file.  Must be a valid\n",
      "                        character (alphanumeric, period, or\n",
      "                        underscore).[default: _]\n",
      "  -b, --not_barcoded    Use -b if barcodes are not present.  BarcodeSequence\n",
      "                        header still required.  [default: False]\n",
      "  -B, --variable_len_barcodes\n",
      "                        Use -B if variable length barcodes are present to\n",
      "                        suppress warnings about barcodes of unequal length.\n",
      "                        [default: False]\n",
      "  -p, --disable_primer_check\n",
      "                        Use -p to disable checks for primers.\n",
      "                        LinkerPrimerSequence header still required. [default:\n",
      "                        False]\n",
      "  -j ADDED_DEMULTIPLEX_FIELD, --added_demultiplex_field=ADDED_DEMULTIPLEX_FIELD\n",
      "                        Use -j to add a field to use in the mapping file as\n",
      "                        additional demultiplexing (can be used with or without\n",
      "                        barcodes).  All combinations of barcodes/primers and\n",
      "                        the these fields must be unique. The fields must\n",
      "                        contain values that can be parsed from the fasta\n",
      "                        labels such as \"plate=R_2008_12_09\".  In this case,\n",
      "                        \"plate\" would be the column header and \"R_2008_12_09\"\n",
      "                        would be the field data (minus quotes) in the mapping\n",
      "                        file.  To use the run prefix from the fasta label,\n",
      "                        such as \">FLP3FBN01ELBSX\", where \"FLP3FBN01\" is\n",
      "                        generated from the run ID, use \"-j run_prefix\" and set\n",
      "                        the run prefix to be used as the data under the column\n",
      "                        header \"run_prefix\".  [default: none]\n",
      "  -s, --suppress_html   Use -s to disable html file generation, can be useful\n",
      "                        for extremely large mapping files. [default: False]\n",
      "\n",
      "  REQUIRED options:\n",
      "    The following options must be provided under all circumstances.\n",
      "\n",
      "    -m MAPPING_FP, --mapping_fp=MAPPING_FP\n",
      "                        Metadata mapping filepath [REQUIRED]\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# activate conda environment\n",
    "source activate py2\n",
    "validate_mapping_file.py -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### Validating (actuall call)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2017-09-14T18:46:53.132Z"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "source activate py2\n",
    "# qiime: validate mapping file script\n",
    "validate_mapping_file.py \\\n",
    "    -m mapping_file.txt \\\n",
    "    -o validate_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2017-09-14T18:46:53.564Z"
    }
   },
   "outputs": [],
   "source": [
    "ls -thlc validate_map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "### View html output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2017-09-14T18:46:57.244Z"
    }
   },
   "outputs": [],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "\n",
    "with open('validate_map/mapping_file.html') as f:\n",
    "    output = f.read()\n",
    "display(HTML(output))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "kernelspec": {
   "display_name": "Python [conda env:py2]",
   "language": "python",
   "name": "conda-env-py2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
